courier-imap-x/libs/maildir/maildirmake.1.in
'\" t
.\"<!-- Copyright 1998 - 2018 Double Precision, Inc.  See COPYING for -->
.\"<!-- distribution information. -->
.\"     Title: maildirmake
.\"    Author: Sam Varshavchik
.\" Generator: DocBook XSL Stylesheets vsnapshot <http://docbook.sf.net/>
.\"      Date: 04/14/2022
.\"    Manual: Double Precision, Inc.
.\"    Source: Courier Mail Server
.\"  Language: English
.\"
.TH "MAILDIRMAKE" "1" "04/14/2022" "Courier Mail Server" "Double Precision, Inc\&."
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
maildirmake \- create maildirs and maildir folders
.SH "SYNOPSIS"
.HP \w'\fBmaildirmake\fR\ 'u
\fBmaildirmake\fR [options...] {\fImaildir\fR}
.SH "DESCRIPTION"
.PP
The
\fBmaildirmake\fR
command creates maildirs, and maildir folders and performs some routine maintenance on them\&. This documentation describes the
\fBmaildirmake\fR
command from the
Courier
mail server, which creates an extended form of maildirs that implements additional extensions beyond the basic maildir properties that were first implemented in the Qmail mail server\&.
.SH "OPTIONS"
.PP
\-S
.RS 4
create a "sharable" maildir\&. A sharable maildir has slightly different permissions which allows creation of publicly\-shared folders\&.
.RE
.PP
\-q \fIquota\fR
.RS 4
install a quota on the maildir\&. See
\m[blue]\fB\fBmaildirquota\fR(7)\fR\m[]\&\s-2\u[1]\d\s+2
for more information\&. The specified maildir gets automatically created if it does not exist; otherwise the existing maildir\*(Aqs quota gets updated\&.
\fIquota\fR
may be:
.PP
\fI####S\fR
.RS 4
Specifies the quota in the total number of bytes for all messages\&.
\(lq10000000S\(rq
specifies a quota of ten million bytes\&.
.RE
.PP
\fI####C\fR
.RS 4
Specifies the quota in the total number of messages in the maildir\&.
\(lq10000S\(rq
specifies a quota of ten thousand messages\&.
.RE
.RE
.PP
\-f \fIfolder\fR
.RS 4
do not create a maildir, but create a folder in an existing maildir\&.
.RE
.PP
\-F \fIfolder\fR
.RS 4
Like the
\-f
option, except that the folder\*(Aqs name is given using the system locale\*(Aqs character set\&. Non\-Latin characters in the folder\*(Aqs name must be given to the
\-f
option using IMAP\*(Aqs UTF8 encoding\&. The
\-F
option takes the folder name specified using the console\*(Aqs character set\&.
.RE
.PP
\-s \fImode\fR
.RS 4
create a publicly accessible folder in an existing sharable maildir\&. First, use the
\fB\-S\fR
option to create a sharable maildir\&. Then, run
\fBmaildirmake\fR
again with the
\fB\-s\fR
option to create publicly accessible folders\&.
\fImode\fR
is a comma\-separated list of the following keywords:
read
\- readonly folder, only you can write messages to this folder;
write
\- anyone can read and write messages to this folder;
group
\- only allow members of your own system group to access messages in this folder (instead of everyone)\&.
.RE
.PP
\-\-add \fIname\fR=\fIpathname\fR, \-\-del \fIname\fR
.RS 4
create or delete the directories and links needed to access shared folders\&. See below for more information\&.
.RE
.PP
\-\-checkutf8 \fImaildir\fR \fImaildirfilter\fR
.RS 4
Perform a sanity check to verify that a pre\-unicode format maildir can be converted to a unicode\-format maildir\&. See
\(lqConverting pre\-unicode format maildirs\(rq, below, for more information\&.
.RE
.PP
\-\-convutf8 \fImaildir\fR \fImaildirfilter\fR
.RS 4
Convert a pre\-unicode format maildir can be converted to a unicode\-format maildir\&. See
\(lqConverting pre\-unicode format maildirs\(rq, below, for more information\&.
.RE
.SS "FOLDERS"
.PP
This
\fBmaildirmake\fR
command supports enhanced maildirs that contain folders\&.
.PP
By itself,
\fBmaildirmake\fR
makes a new subdirectory
\fImaildir\fR, and creates all the necessary structures\&. The
\fB\-f\fR
option creates a new "folder" within an existing
\fImaildir\fR\&.
\fImaildir\fR
must already exist, and the
\fBmaildirmake\fR
command will create a new folder in the maildir\&.
.PP
Folders are simply subdirectories inside the main maildir whose names start with a period, and which are themselves maildirs\&. For example, the command "\fBmaildirmake \-f Drafts mail/Maildir\fR" creates
mail/Maildir/\&.Drafts, that has the usual
tmp,
new
and
cur\&. You MUST use the
\fB\-f\fR
option, instead of specifying
mail/Maildir/\&.Drafts
directly, in order to correctly initialize certain enhanced maildir features\&.
.PP
Folders cannot be created directly within other folders\&. Running
\fBmaildirmake \-f Urgent mail/Maildir/\&.Drafts\fR
will not work\&. Instead, the period character is designated as a hierarchy separator, run
\fBmaildirmake \-f Drafts\&.Urgent mail/Maildir\fR
instead\&. This creates
mail/Maildir/\&.Drafts\&.Urgent, and all mail software that supports enhanced maildirs will interpret it as a subfolder Urgent of the Drafts folder\&.
.SS "SHARED FOLDERS"
.PP
This is another extension to the Maildir format that allows folders to be shared between multiple clients\&.
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
.PP
The
Courier
IMAP server implements two types of shared folders: filesystem permission\-based shared folders, as well as virtual shared folders based on IMAP access control lists\&. Use the
\fBmaildirmake\fR
command to implement shared folders based on filesystem permissions\&. The
\m[blue]\fB\fBmaildiracl\fR(1)\fR\m[]\&\s-2\u[2]\d\s+2
command manages access control lists, which are used by virtual shared folders\&.
.PP
See the
Courier
IMAP server documentation for more information\&.
.sp .5v
.RE
.PP
First, you need to create a collection of sharable folders, as a separate maildir:
.sp
.if n \{\
.RS 4
.\}
.nf
\fBmaildirmake \-S /usr/local/share/maildirs/notices\fR
.fi
.if n \{\
.RE
.\}
.PP
Then, create individuals folders that will be accessed in shared mode:
.sp
.if n \{\
.RS 4
.\}
.nf
\fBmaildirmake \-s write \-f Weekly /usr/local/share/maildirs/notices\fR
.fi
.if n \{\
.RE
.\}
.PP
In this example, the "Weekly" folder is created, with read/write access to everyone\&. Multiple folders can be created in the same maildir, with different access permissions\&. Everyone can create a sharable maildir\&. The access privileges for individual folders are set by the
\fB\-s\fR
option, and are implemented using traditional filesystem permissions\&.
.PP
Use the
\fB\-\-add\fR
and
\fB\-\-del\fR
options to add a sharable maildir to an existing maildir\&. Client software that implements this extension will now know where to find sharable folders:
.sp
.if n \{\
.RS 4
.\}
.nf
\fBmaildirmake \-\-add notices=/usr/local/share/maildirs/notices $HOME/Maildir\fR
.fi
.if n \{\
.RE
.\}
.PP
$HOME/Maildir
is your main maildir\&. The argument to
\fB\-add\fR
is
\fInick\fR=\fIpath\fR\&.
\fInick\fR
is a nickname for this collection of sharable folders, and
\fIpath\fR
is the location of the sharable maildir\&. All folders in the sharable maildir that you have access to \-\- such as "Weekly", in this case, will now be accessible\&. Multiple sharable maildirs can be added, by giving each one a unique
\fInick\fR\&.
.PP
The
\fB\-\-del\fR
option "disconnects" the sharable maildir from the main maildir\&.
.SS "GLOBAL SHARED FOLDERS"
.PP
Normally
\fB\-add\fR
command must be run for every maildir which needs to access the sharable maildir\&. Alternatively the file
@sysconfdir@/maildirshared
can be created, to specify a default set of sharable maildirs\&. Each line in this file takes the following format:
.sp
.if n \{\
.RS 4
.\}
.nf
\fInick\fR<tab>\fIpath\fR
.fi
.if n \{\
.RE
.\}
.PP
\fInick\fR
is a short nickname for the sharable maildir,
<tab>
is a single tab character,
\fIpath\fR
is the pathname to the sharable maildir\&.
.SS "ACCESSING SHARED FOLDERS"
.PP
You may have read or write access to a shared folder\&. If you have write access, you can add messages to the shared folder\&. You can also delete messages that you\*(Aqve added\&.
.PP
Anyone can create a sharable maildir, so if the sharable maildir is actually created by you, can can delete any message, not just your own\&.
.SH "CONVERTING PRE\-UNICODE FORMAT MAILDIRS"
.PP
This section is relevant to:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Updating
Courier\-IMAP
to version 5\&.0, and later, from prior versions of
Courier\-IMAP, or:
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Updating
SqWebmail
to version 6\&.0, and later, from prior versions of
SqWebmail, or:
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Updating Courier to version 1\&.0, and later, from prior versions of Courier\&.
.RE
.PP
These versions have been updated to implement native Unicode support in several E\-mail\-related protocols\&. It is already expected that updating Internet standards to use native Unicode\-formatted E\-mail messages will not be 100% backwards\-compatible, in terms of E\-mail client support\&. Given that, this major update to Unicode will also introduce some backwards\-incompatible changes to the internal structure of maildirs, as a major upgrade to simplify Unicode support going forward\&. Might as well go through the pain of a major upgrade once\&.
.PP
\fBmaildirmake\fR\*(Aqs
\fB\-\-checkutf8\fR
and
\fB\-\-convutf8\fR
options are tools to aid in conversion of existing mailboxes to the new Unicode\-based naming standard\&.
.SS "Background"
.PP
Mail folders in a maildir are hidden subdirectories\&. For example: a folder name
\(lqMailing list\(rq
is a maildir subdirectory named
$HOME/Maildir/\&.Mailing list
($HOME/Maildir
is the main mailbox)\&.
.PP
Prior to the unicode update, non\-English characters in folder names used a convention based on the non\-standard
\(lqmodified\-UTF7\(rq
encoding used by IMAP\&. A folder named
\(lqRésumé\(rq
is a maildir subdirectory named
$HOME/Maildir/\&.R&AOk\-sum&AOk\-\&. The current versions of Courier,
Courier\-IMAP, and SqWebmail, now creates
$HOME/Maildir/\&.Résumé
using the
UTF8
encoding\&. This appears as plain
\(lq\&.Résumé\(rq
(hidden) subdirectory on modern UTF8\-based systems\&.
.PP
Consequently, any existing maildirs with folders that use non\-English names must be converted as part of updating to the current version of Courier,
Courier\-IMAP, and SqWebmail from pre\-unicode versions\&. This does not happen automatically when updating to the current version\&. This must be done manually given the wide variety of individual mail server configurations that are possible\&.
.SS "Unicode conversion overview"
.PP
Updating from pre\-unicode versions involves:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Renaming the actual maildir folders,
$HOME/Maildir/\&.\fInames\fR
into unicode names (using
UTF8)\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Updating the
$HOME/Maildir/courierimapsubscribed
file, which is a list of subscribed IMAP folders, if it exists\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Updating any
maildrop
mail filtering recipes,
$HOME/\&.mailfilter, if it exists, to reference the unicode maildir folders; or updating any custom site mail filtering engine that delivers to maildir folders, to reference the correct subdirectory names\&.
.RE
.SS "Unicode conversion steps"
.PP
The
\fB\-\-checkutf8\fR
and
\fB\-\-convutf8\fR
options to
\fBmaildirmake\fR
convert a single maildir to the new unicode format:
.sp
.if n \{\
.RS 4
.\}
.nf
$ \&./maildirmake \-\-checkutf8 ~/Maildir ~/\&.mailfilter
Checking /home/mrsam/Maildir:
Mail filter to INBOX\&.R&AOk\-sum&AOk\- updated to INBOX\&.Résumé
Subscription to INBOX\&.R&AOk\-sum&AOk\- changed to INBOX\&.Résumé
Rename INBOX\&.R&AOk\-sum&AOk\- to INBOX\&.Résumé
Verified /home/mrsam/Maildir/courierimapsubscribed
Verified /home/mrsam/\&.mailfilter
$ \&./maildirmake \-\-convutf8 ~/Maildir ~/\&.mailfilter
Checking /home/mrsam/Maildir:
Mail filter to INBOX\&.R&AOk\-sum&AOk\- updated to INBOX\&.Résumé
Subscription to INBOX\&.R&AOk\-sum&AOk\- changed to INBOX\&.Résumé
Rename INBOX\&.R&AOk\-sum&AOk\- to INBOX\&.Résumé
Updating /home/mrsam/Maildir/courierimapsubscribed
Updating /home/mrsam/\&.mailfilter
.fi
.if n \{\
.RE
.\}
.PP
\fB\-\-checkutf8\fR
goes through the motions of converting a single maildir to Unicode, but without making any actual changes\&.
\fB\-\-convutf8\fR
does the conversion for real\&. The first required parameter is the maildir to convert\&. The second parameter is optional, and specifies the corresponding
\fBmaildrop\fR
filtering recipe,
\fIbut only if \fR\fISqWebMail\fR
generates the mail filtering recipes\&.
SqWebMail\*(Aqs mail filtering recipes are parsable, and can be automatically\-converted\&. Non\-SqWebMail\-generated
\&.mailfilters cannot be converted automatically\&. The second parameter must be omitted, and the mail filtering recipe must be converted by hand\&.
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
.PP
All this work is only needed if maildirs have folders with non\-English names\&. Ignore everything you\*(Aqve just read if all folder names are English\-only\&.
\fB\-\-checkutf8\fR
and
\fB\-\-convutf8\fR
will not do anything, and nothing needs to be done\&.
.sp .5v
.RE
.PP

indimail-x/NEWS
doc/RELEASE-Notes-indimail.md

courier-imap-x/libs/maildir/maildirmake.c
/*
** Copyright 1998 - 2018 Double Precision, Inc.
** See COPYING for distribution information.
*/


#if	HAVE_CONFIG_H
#include	"config.h"
#endif
#include	<stdio.h>
#include	<string.h>
#include	<stdlib.h>
#if	HAVE_FCNTL_H
#include	<fcntl.h>
#endif
#if	HAVE_UNISTD_H
#include	<unistd.h>
#endif
#include	<sys/types.h>
#if	HAVE_SYS_STAT_H
#include	<sys/stat.h>
#endif
#include	<ctype.h>
#include	<errno.h>

#include	"maildircreate.h"
#include	"maildirmisc.h"
#include	"maildirinfo.h"
#include	"maildirsharedrc.h"
#include	"maildirquota.h"
#include	"maildirfilter.h"
#include	<unicode/courier-unicode.h>

static void usage()
{
	printf("Usage: maildirmake [ options ] maildir\n");
	exit(1);
}

extern FILE *maildir_shared_fopen(const char *, const char *);
extern void maildir_shared_fparse(char *, char **, char **);

static void add(const char *dir, const char *name)
{
char	*c=malloc(strlen(name)+1);
char	*s;
FILE	*fp;
char	buf[BUFSIZ];
char	*ptr, *dirp;

	if (!c)
	{
		perror("malloc");
		exit(1);
	}
	strcpy(c, name);

	if ((s=strchr(c, '=')) == 0)
		usage();
	*s++=0;
	if (*s != '/')
		usage();
	if (access(s, R_OK))
	{
		perror(s);
		exit(1);
	}
	if (strchr(c, '.') || strchr(c, '/'))
	{
		fprintf(stderr, "%s: invalid name\n", c);
		exit(1);
	}

	if (chdir(dir))
	{
		perror(dir);
		exit(1);
	}

	if ((fp=fopen(MAILDIRSHAREDRC, "r")) != 0)
	{
		while ((ptr=fgets(buf, sizeof(buf), fp)) != 0)
		{
		char	*namep;

			maildir_shared_fparse(buf, &namep, &dirp);
			if (!namep)	continue;
			if (strcmp(namep, c) == 0)
			{
				fclose(fp);
				fprintf(stderr,
		"%s: already defined as a sharable maildir in %s.\n",
					namep, MAILDIRSHAREDRC);
				exit(2);
			}
		}
		fclose(fp);
	}

	if ((fp=maildir_shared_fopen(".", "a+")) == 0)
	{
		perror(dir);
		exit(1);
	}

	if (fseek(fp, 0L, SEEK_SET) < 0)
	{
		perror(dir);
		exit(1);
	}

	while ((ptr=fgets(buf, sizeof(buf), fp)) != 0)
	{
	char	*namep;

		maildir_shared_fparse(buf, &namep, &dirp);
		if (!namep)	continue;
		if (strcmp(namep, c) == 0)
		{
			fclose(fp);
			fprintf(stderr, "%s: already defined as a sharable maildir.\n",
				namep);
			exit(2);
		}
	}
	fprintf(fp, "%s\t%s\n", c, s);
	if (fflush(fp) || ferror(fp) || fclose(fp))
	{
		perror("dir");
		exit(1);
	}
	exit(0);
}

static void del(const char *dir, const char *n)
{
FILE	*fp;
char	buf[BUFSIZ];
char	buf2[BUFSIZ];

FILE	*fp2;
int	found;
 struct maildir_tmpcreate_info createInfo;

	if (chdir(dir))
	{
		perror(dir);
		exit(1);
	}
	if ((fp=maildir_shared_fopen(".", "r")) == 0)
	{
		perror(dir);
		exit(1);
	}

	maildir_tmpcreate_init(&createInfo);

	createInfo.uniq="shared";
	createInfo.msgsize=0;
	createInfo.doordie=1;

	if ((fp2=maildir_tmpcreate_fp(&createInfo)) == NULL)
	{
		perror(dir);
		exit(1);
	}

	found=0;
	while (fgets(buf, sizeof(buf), fp))
	{
	char	*name, *dirp;

		strcpy(buf2, buf);
		maildir_shared_fparse(buf2, &name, &dirp);
		if (name && strcmp(name, n) == 0)
		{
			found=1;
			continue;
		}
		fprintf(fp2, "%s", buf);
	}
	fclose(fp);
	if (fflush(fp2) || ferror(fp2) || fclose(fp2)
	    || rename(createInfo.tmpname, "shared-maildirs"))
	{
		perror(createInfo.tmpname);
		unlink(createInfo.tmpname);
		exit(1);
	}
	maildir_tmpcreate_free(&createInfo);
	if (!found)
	{
		fprintf(stderr, "%s: not found.\n", n);
		exit(1);
	}
	exit(0);
}

/*****************************************************************************

Convert modified-UTF7 folder names to UTF-8 (sort-of).

*****************************************************************************/

struct convertutf8_list {
	struct convertutf8_list *next;
	char *rename_from;
	char *rename_to;
};

struct convertutf8_status {
	struct convertutf8_list *list;
	int status;
};

/* Find folders that need to change */

static void convertutf8_build_list(const char *inbox_name,
				   void *arg)
{
	struct convertutf8_status *status=
		(struct convertutf8_status *)arg;
	struct convertutf8_list *list;

	char *converted=imap_foldername_to_filename(0, inbox_name);

	if (!converted)
	{
		fprintf(stderr,
			"Error: %s: does not appear to be valid"
			" modified-UTF7\n",
			inbox_name);
		status->status=1;
		return;
	}

	if (strcmp(converted, inbox_name) == 0)
	{
		free(converted);
		return;
	}
	list=(struct convertutf8_list *)malloc(sizeof(struct convertutf8_list));

	if (!list || !(list->rename_from=strdup(inbox_name)))
	{
		perror("malloc");
		exit(1);
	}

	list->rename_to=converted;
	list->next=status->list;
	status->list=list;
}

void convertutf8(const char *maildir, const char *mailfilter, int doit)
{
	struct convertutf8_status status;
	struct convertutf8_list *list;
	char *courierimapsubscribed;
	char *courierimapsubscribed_new;
	FILE *courierimapsubscribed_fp;
	struct maildirfilter mf;
	int mf_status;
	char *mailfilter_newname=0;

	memset(&status, 0, sizeof(status));
	memset(&mf, 0, sizeof(mf));

	printf("Checking %s:\n", maildir);

	maildir_list(maildir, convertutf8_build_list, &status);

	if (status.status)
		exit(status.status);

	if (mailfilter)
	{
		struct maildirfilterrule *r;

		/*
		** Try to convert folder references from mailfilter
		*/

		mailfilter_newname=malloc(strlen(mailfilter)+10);

		strcat(strcpy(mailfilter_newname, mailfilter), ".new");
		mf_status=maildir_filter_loadrules(&mf, mailfilter);

		if (mf_status != MF_LOADOK && mf_status != MF_LOADNOTFOUND)
		{
			fprintf(stderr, "Error: cannot load %s\n",
				mailfilter);
		}

		for (r=mf.first; r; r=r->next)
		{
			char *converted;

			/* Look for deliveries to a folder */

			if (strncmp(r->tofolder, "INBOX.", 6))
				continue;

			converted=imap_foldername_to_filename(0, r->tofolder);

			if (!converted)
			{
				fprintf(stderr, "Error: %s: "
					"%s: does not appear to be valid "
					"modified-UTF7\n",
					mailfilter,
					r->tofolder);
				status.status=1;
			}

			if (strcmp(converted, r->tofolder) == 0)
			{
				free(converted);
				continue;
			}

			printf("Mail filter to %s updated to %s\n",
			       r->tofolder, converted);

			free(r->tofolder);
			r->tofolder=converted;
		}

		if (mf_status == MF_LOADOK)
		{
			FILE *fp=fopen(mailfilter, "r");
			char detected_from[1024];
			char detected_tomaildir[1024];
			char buffer[1024];
			struct stat st_buf;

			/*
			** We need to know the FROM address and the MAILDIR
			** reference. Attempt to ham-fistedly parse the
			** current filter file.
			*/

			detected_from[0]=0;
			detected_tomaildir[0]=0;

			if (!fp)
			{
				perror(mailfilter);
				exit(1);
			}

			while (fgets(buffer, sizeof(buffer), fp))
			{
				char *p, *q;

				if (strncmp(buffer, "FROM='", 6) == 0)
				{
					char *p=strchr(buffer+6, '\'');

					if (!p)
					{
						fprintf(stderr,
							"Cannot parse %s\n",
							mailfilter);
						status.status=1;
					}
					else
					{
						*p=0;
						strcpy(detected_from, buffer+6);

						/*
						  Unescape, because saverules()
						  escapes it.
						*/

						for (q=p=detected_from; *p; p++)
						{
							if (*p == '\\' && p[1])
								++p;
							*q=*p;
							++q;
						}
						*q=0;
					}
				}

				if (strncmp(buffer, "to \"", 4) == 0)
				{
					p=strstr(buffer+4, "/.\"");

					if (!p)
					{
						fprintf(stderr,
							"Cannot parse %s\n",
							mailfilter);
						status.status=1;
					}
					else
					{
						*p=0;
					}
					strcpy(detected_tomaildir, buffer+4);
				}
			}
			fclose(fp);
			if (detected_from[0] == 0 ||
			    detected_tomaildir[0] == 0)
			{
				fprintf(stderr,
					"Failed to parse %s\n",
					mailfilter);
				status.status=1;
			}
			maildir_filter_saverules(&mf, mailfilter_newname,
						 detected_tomaildir,
						 detected_from);

			if (stat(mailfilter, &st_buf) ||
			    chmod(mailfilter_newname, st_buf.st_mode))
			{
				perror(mailfilter);
				exit(1);
			}
			/*
			** If we're root, preserve the ownership and permission
			*/
			if (geteuid() == 0)
			{
				if (chown(mailfilter_newname, st_buf.st_uid,
					  st_buf.st_gid))
				{
					perror(mailfilter_newname);
					exit(1);
				}
			}
		}
	}
	else
	{
		mf_status=MF_LOADNOTFOUND;
	}

	courierimapsubscribed=malloc(strlen(maildir)+100);
	courierimapsubscribed_new=malloc(strlen(maildir)+100);

	strcat(strcpy(courierimapsubscribed, maildir),
	       "/courierimapsubscribed");
	strcat(strcpy(courierimapsubscribed_new, maildir),
	       "/courierimapsubscribed.new");

	/*
	** Update folder references in the IMAP subscription file.
	*/

	if ((courierimapsubscribed_fp=fopen(courierimapsubscribed, "r"))
	    != NULL)
	{
		char buffer[2048];
		FILE *new_fp=fopen(courierimapsubscribed_new, "w");
		char *converted;
		struct stat st_buf;

		while (fgets(buffer, sizeof(buffer), courierimapsubscribed_fp))
		{
			char *p=strchr(buffer, '\n');

			if (!p)
			{
				fprintf(stderr, "Error: courierimapsubscribed: "
					"folder name too long\n");
				status.status=1;
				continue;
			}

			*p=0;

			converted=imap_foldername_to_filename(0, buffer);

			if (!converted)
			{
				fprintf(stderr, "Error: courierimapsubscribed: "
					"%s: does not appear to be valid "
					"modified-UTF7\n",
					buffer);
				status.status=1;
				continue;
			}
			fprintf(new_fp, "%s\n", converted);

			if (strcmp(buffer, converted))
			{
				printf("Subscription to %s changed to %s\n",
				       buffer, converted);
			}
			free(converted);
		}
		if (fflush(new_fp) || fclose(new_fp))
		{
			exit(1);
		}
		fclose(courierimapsubscribed_fp);

		/*
		** If we're root, preserve the ownership and permission
		*/
		if (stat(courierimapsubscribed, &st_buf) ||
		    chmod(courierimapsubscribed_new, st_buf.st_mode))
		{
			perror(courierimapsubscribed);
			exit(1);
		}

		if (geteuid() == 0)
		{
			if (chown(courierimapsubscribed_new, st_buf.st_uid,
				  st_buf.st_gid))
			{
				perror(courierimapsubscribed_new);
				exit(1);
			}
		}
	}

	if (status.status)
	{
		unlink(courierimapsubscribed_new);
		if (mf_status == MF_LOADOK)
			unlink(mailfilter_newname);
		exit(status.status);
	}
	for (list=status.list; list; list=list->next)
	{
		char *frompath, *topath;

		printf("Rename %s to %s\n", list->rename_from, list->rename_to);

		frompath=malloc(strlen(maildir)+strlen(list->rename_from));
		topath=malloc(strlen(maildir)+strlen(list->rename_to));

		if (!frompath || !topath)
		{
			perror("malloc");
			exit(1);
		}

		strcat(strcpy(frompath, maildir), "/");
		strcat(strcpy(topath, maildir), "/");

		/* They all have the INBOX. prefix, strip it off */
		strcat(frompath, strchr(list->rename_from, '.'));
		strcat(topath, strchr(list->rename_to, '.'));

		if (doit)
		{
			if (rename(frompath, topath))
			{
				fprintf(stderr,
					"FATAL ERROR RENAMING %s to %s: %s\n",
					frompath, topath, strerror(errno));
				status.status=1;
			}
		}
		free(frompath);
		free(topath);
	}

	if (doit)
	{
		if (courierimapsubscribed_fp)
		{
			printf("Updating %s\n", courierimapsubscribed);

			if (rename(courierimapsubscribed_new,
				   courierimapsubscribed))
			{
				fprintf(stderr,
					"FATAL ERROR RENAMING %s to %s: %s\n",
					courierimapsubscribed_new,
					courierimapsubscribed, strerror(errno));
				status.status=1;
			}
		}

		if (mf_status == MF_LOADOK)
		{
			printf("Updating %s\n", mailfilter);

			if (rename(mailfilter_newname, mailfilter))
			{
				fprintf(stderr,
					"FATAL ERROR RENAMING %s to %s: %s\n",
					mailfilter_newname, mailfilter,
					strerror(errno));
				status.status=1;
			}
		}
	}
	else
	{
		if (courierimapsubscribed_fp)
		{
			printf("Verified %s\n", courierimapsubscribed);
		}

		if (mf_status == MF_LOADOK)
		{
			printf("Verified %s\n", mailfilter);
		}
	}
	exit(status.status);
}

int main(int argc, char *argv[])
{
const char *maildir, *folder=0;
char *p;
int	argn;
int	perm=0700;
int	musthavefolder=0;
int	subdirperm;
char	*addshared=0, *delshared=0;
const char *quota=0;
char	*tmp=0;

	for (argn=1; argn<argc; argn++)
	{
		if (argv[argn][0] != '-')	break;
		if (strcmp(argv[argn], "-") == 0)	break;
		if (strncmp(argv[argn], "-f", 2) == 0)
		{
			folder=argv[argn]+2;
			if (*folder == 0 && argn+1 < argc)
				folder=argv[++argn];
			continue;
		}
		if (strncmp(argv[argn], "-F", 2) == 0)
		{
			int converr;

			const char *p=argv[argn]+2;

			if (*p == 0 && argn+1 < argc)
				p=argv[++argn];

			folder=unicode_convert_tobuf(p,
						       unicode_default_chset(),
						       unicode_x_smap_modutf8,
						       &converr);

			if (converr || !folder)
			{
				fprintf(stderr, "Cannot convert %s to maildir encoding\n",
					p);
				exit(1);
			}
			continue;
		}
		if (strcmp(argv[argn], "-S") == 0)
		{
			perm=0755;
			continue;
		}

		if (strncmp(argv[argn], "-s", 2) == 0)
		{
		const char *p=argv[argn]+2;

			if (*p == 0 && argn+1 < argc)
				p=argv[++argn];

			perm=0755;
			for (; *p; ++p)
			{
				if (isspace((int)(unsigned char)*p) ||
					*p == ',')
					continue;
				if (*p == 'r')
					perm=0755;
				else if (*p == 'w')
					perm=0777 | S_ISVTX;
				else if (*p == 'g')
					perm &= ~0007;

				while (*p && !isspace((int)(unsigned char)*p)
					&& *p != ',')
					++p;
				if (!*p)	break;
			}
			musthavefolder=1;
			continue;
		}

		if (strncmp(argv[argn], "-q", 2) == 0)
		{
			const char *p=argv[argn]+2;

			if (*p == 0 && argn+1 < argc)
				p=argv[++argn];

			quota=p;
			continue;
		}

		if (strcmp(argv[argn], "--add") == 0 && argc-argn > 1)
		{
			addshared=argv[++argn];
			continue;
		}

		if (strcmp(argv[argn], "--del") == 0 && argc-argn > 1)
		{
			delshared=argv[++argn];
			continue;
		}

		if (strcmp(argv[argn], "--convutf8") == 0 && argc-argn > 1)
		{
			char *maildir=argv[++argn];
			char *mailfilter=argn < argc ? argv[++argn]:0;
			convertutf8(maildir, mailfilter, 1);
			exit(0);
		}

		if (strcmp(argv[argn], "--checkutf8") == 0 && argc-argn > 1)
		{
			char *maildir=argv[++argn];
			char *mailfilter=argn < argc ? argv[++argn]:0;
			convertutf8(maildir, mailfilter, 0);
			exit(0);
		}
		usage();
	}

	if (argn == argc)	usage();
	maildir=argv[argn];

	if (addshared)
	{
		add(maildir, addshared);
		exit (0);
	}

	if (delshared)
	{
		del(maildir, delshared);
		exit (0);
	}

	if (folder && *folder == '.')
	{
		printf("Invalid folder name: %s\n", folder);
		exit(1);
	}

	if (folder)
	{
		if ((p=(char *)malloc(strlen(maildir)+strlen(folder)+4)) == 0)
		{
			perror("maildirmake malloc");
			exit(1);
		}
		strcat(strcat(strcpy(p, maildir), "/."), folder);
		maildir=p;
	}
	else	if (musthavefolder)
		usage();

	if (quota)
	{
		struct stat stat_buf;

		if (stat(maildir, &stat_buf) < 0 && errno == ENOENT)
		{
			if (maildir_make(maildir, perm & ~0022,
					 (perm & ~0022),
					 0) < 0)
			{
				perror(maildir);
				exit(1);
			}
		}

		maildir_quota_set(maildir, quota);
		exit(0);
	}
	subdirperm=perm;
	if (!folder)	subdirperm=0700;
	umask(0);
	if (maildir_make(maildir, perm & ~0022, subdirperm, folder ? 1:0) < 0)
	{
		tmp=(char *)malloc(strlen(maildir)+14);
		if (!tmp) {
			perror("maildirmake malloc");
			exit(1);
		}
		snprintf(tmp, 14+strlen(maildir), "maildirmake: %s", maildir);
		perror(tmp);
		exit(1);
	}
	exit(0);
	return (0);
}
